WordPress/Classes/ViewRelated/Media/AnimatedImageCache.swift

```diff
@@ -23,50 +19,38 @@ class AnimatedImageCache {
         return session
     }()
 
-    fileprivate static var cache: NSCache<AnyObject, AnyObject> = {
-        let cache = NSCache<AnyObject, AnyObject>()
-        cache.totalCostLimit = totalCostLimit()
-
-        let bcf = ByteCountFormatter()
-        bcf.allowedUnits = [.useMB]
-        bcf.countStyle = .file
-        let cacheSizeinMB = bcf.string(fromByteCount: Int64(totalCostLimit()))
-
-        return cache
-    }()
-
     // MARK: Instance methods
 
-    @objc func handleMemoryWarning() {
-        clearCache()
-    }
-
-    func clearCache() {
-        AnimatedImageCache.cache.removeAllObjects()
+    func cacheData(data: Data, url: URL?) {
+        guard let url = url else {
+            return
+        }
+        let key = url.absoluteString + Constants.keyDataSuffix
+        MemoryCache.shared.setData(data, forKey: key)
     }
 
     func cachedData(url: URL?) -> Data? {
-        guard let key = url else {
+        guard let url = url else {
             return nil
         }
-        return AnimatedImageCache.cache.object(forKey: key as AnyObject) as? Data
+        let key = url.absoluteString + Constants.keyDataSuffix
+        return MemoryCache.shared.geData(forKey: key)
     }
 
     func cacheStaticImage(url: URL?, image: UIImage?) {
-        guard let url = url,
-            let image = image else {
-                return
+        guard let url = url, let image = image else {
+            return
         }
         let key = url.absoluteString + Constants.keyStaticImageSuffix
-        AnimatedImageCache.cache.setObject(image as AnyObject, forKey: key as AnyObject)
+        MemoryCache.shared.setImage(image, forKey: key)
```

Comment:
> Does it mean that there's a high chance of many images being removed from the cache after encountering a few heavy GIFs, given we're limiting the cache to 256 MB?

Yes, because the alternative would be overusing memory and potentially risking app termination. The limits are imprecise, but I think 256 MB may be a bit low. I would rather start with a conservative limit to prioritize eliminating the crashes.

Getting the best possible experience with large images like that will require more work. I would prioritize the following:

- Set a maximum cost limit for individual entries. Let's say it's larger than 10% of the total cost limit – prevent it from being stored in memory.
- Memory cache is secondary. The primary cache should be the (fast) disk cache. Instead of trying to cram as much as possible into the memory cache, the app should prefetch the images as you scroll to make sure it _appears_ as if they are cached in memory. In many situations, it will appear that even the images fetched from the network are instantly available.
- Stop using GIFs. MP4 is often 100 [times smaller](https://web.dev/replace-gifs-with-videos/) than GIF, has better quality, and uses literally 100 times less CPU and memory to render.
- Downscale large images to make sure they take less space in memory.

I outlined some other potential optimizations [here](https://docs.google.com/spreadsheets/d/1ivWwq5PfnPRAAbv5UaNQoq1oz2gli0bHDDWehyotbJo/edit#gid=0).

Reply:
Got it, thanks for the explanation!

> Yes, because the alternative would be overusing memory and potentially risking app termination. The limits are imprecise, but I think 256 MB may be a bit low. But I would rather start with a conservative limit to prioritize eliminating the crashes.

Agree, and as I noticed testing and [according to the docs](https://developer.apple.com/documentation/foundation/nscache/1407672-totalcostlimit) the limit is not strict and the eviction may not happen instantly if ever. So I guess a lower limit does little harm, especially if the system determines that there's enough memory for the app.

> I outlined some other potential optimizations [here](https://docs.google.com/spreadsheets/d/1ivWwq5PfnPRAAbv5UaNQoq1oz2gli0bHDDWehyotbJo/edit#gid=0).

Thanks for outlying further improvements. I agree that for now we can take it simple and prioritize limiting potential memory overuse and risking app termination.